frontend/app/actions/ActionRefiners.scala

```diff
@@ -33,20 +42,73 @@ object Functions extends LazyLogging {
+    implicit val pf = ProductFamily.membership
+    val tp = request.touchpointBackend
+
+    (for {
+      member <- OptionT(request.forMemberOpt(identity))
+      subscription <- OptionT(tp.subscriptionService.get(member))
+      memberSubscriber <- OptionT(Future.successful(Subscriber.Member.unapply(Subscriber(subscription, member))))
+    } yield new SubscriptionRequest[A](tp, request) with Subscriber {
+      override def subscriber = memberSubscriber
+    }).run
+  }
+
+  def subscriptionRefiner(onNonMember: RequestHeader => Result = notYetAMemberOn(_)) = new ActionRefiner[AuthRequest, SubReqWithSub] {
+    override def refine[A](request: AuthRequest[A]): SubRequestOrResult[A] = {
+      getSubRequest(request).map(_ toRight onNonMember(request))
+    }
+  }
+
+  def paidSubscriptionRefiner(onFreeMember: RequestHeader => Result = notYetAMemberOn(_)) = new ActionRefiner[SubReqWithSub, SubReqWithPaid] {
+    type PaidMemberSub = Sub with PaidPS[PaidMembershipPlan[com.gu.memsub.Status, PaidTier, BillingPeriod]]
+    override protected def refine[A](request: SubReqWithSub[A]): Future[Either[Result, SubReqWithPaid[A]]] =
+      Future.successful((request.subscriber match {
+        case PaidMember(mem) => Some(new SubscriptionRequest[A](request) with PaidSubscriber {
+          override def subscriber = mem
+        })
+        case _ => None
+    }).toRight(onFreeMember(request)))
+  }
+
+
+  def freeSubscriptionRefiner(onPaidMember: RequestHeader => Result = notYetAMemberOn(_)) = new ActionRefiner[SubReqWithSub, SubReqWithFree] {
+    type FreeMemberSub = Sub with FreePS[FreeMembershipPlan[com.gu.memsub.Status, FreeTier]]
+    override protected def refine[A](request: SubReqWithSub[A]): Future[Either[Result, SubReqWithFree[A]]] =
+      Future.successful(request.subscriber match {
+        case FreeMember(mem) => Right(new SubscriptionRequest[A](request) with FreeSubscriber {
+          override def subscriber = mem
+        })
+        case _ => Left(onPaidMember(request))
+      })
+  }
+
+  def checkTierChangeTo(targetTier: PaidTier) = new ActionFilter[SubReqWithSub] {
+    override protected def filter[A](request: SubReqWithSub[A]): Future[Option[Result]] = {
+      if (!request.touchpointBackend.memberService.subscriptionUpgradableTo(request.subscriber.subscription, targetTier)) {
+        Future.successful(Some(Ok(views.html.tier.upgrade.unavailable(request.subscriber.subscription.plan.tier, targetTier))))
+      } else {
+        Future.successful(None)
       }
     }
+  }
 
-  def redirectMemberAttemptingToSignUp(selectedTier: Tier)(req: AnyMemberTierRequest[_]): Result = selectedTier match {
-    case t: PaidTier if t > req.member.tier => tierChangeEnterDetails(t)(req)
+  def redirectMemberAttemptingToSignUp(selectedTier: Tier)(req: SubReqWithSub[_]): Result = selectedTier match {
+    case t: PaidTier if t > req.subscriber.subscription.plan.tier => tierChangeEnterDetails(t)(req)
     case _ => memberHome(req)
```

Comment:
I know this was already there, but we should redirect to `views.html.tier.upgrade.unavailable` instead